Add libdjinterop external project dependency

mr-smidge opened the pull request:
This PR adds the [libdjinterop](https://github.com/xsco/libdjinterop) library as an external project dependency to the CMake build system for Mixxx.  It is a precursor to PR #2753 , which will add library export functionality using libdjinterop.

Notes:
* An existing installation of the library, if available, is used in preference.  If this is not available, a snapshot is downloaded from the libdjinterop GitHub repo.
* The scons build has not been touched, on the assumption that it will be removed soon.
* The library should be considered in beta at present.  As such, the library is added only on Linux builds for now, following prior discussion with devs in PR #2753 .
  * The libdjinterop library has been compiled and tested on 64-bit Windows in isolation, but I haven't integrated this in with Mixxx yet.
  * MacOS builds will fail with the current build servers because Apple LLVM 9 doesn't support `std::optional<T>` (only experimental), which is in the public library headers.

mr-smidge committed:
Add libdjinterop external project dependency

mr-smidge commented on CMakeLists.txt:
```diff
@@ -1438,6 +1438,71 @@ if(WIN32)
+    target_include_directories(mixxx-lib PUBLIC ${DJINTEROP_INCLUDE_DIR})
+    target_link_libraries(mixxx-lib PUBLIC mixxx-libdjinterop)
+
+    # libdjinterop depends on zlib and sqlite3.
+    # Mixxx already has a dependency on sqlite3, but not zlib, which we add now.
+    find_package(ZLIB 1.2.8 REQUIRED)
```
This might be considered slightly fishy: we're linking statically to libdjinterop, but libdjinterop depends on SQLite3 and ZLib.  It just so happens that Mixxx already has the SQLite3 dependency, but not ZLib.

Are there any concerns with encoding implicit knowledge of libdjinterop's dependencies here?  I'm not aware of a better way with  static build.

uklotzde replied:
@Holzhaus Opinions?

Holzhaus replied:
I think this should be move to the FindDJInterop module. For the external project build, this can be taken care of in libdjinterop's own CMakeLists.txt.

mr-smidge replied:
> For the external project build, this can be taken care of in libdjinterop's own CMakeLists.txt.

The [CMakeLists.txt file from libdjinterop](https://github.com/xsco/libdjinterop/blob/6b73cd37bf63f0a1fdbba1652cf8b696f4e38572/CMakeLists.txt#L19) already has a `find_package(ZLIB 1.2.8 REQUIRED)` line.  However, I was not able to get the dependencies of the external project to transfer through to `mixxx-lib` (and search engines were not fruitful for me).  Do you know of a way in CMake to do this?

Technically, the line isn't needed in the `FindDjInterop` because it would link against the shared library, which already has the ZLib dependency.  At the very least (if deps can't be inherited), this is only needed for the static build.

mr-smidge replied:
I just noticed that the ChromaPrint dependency already does this (with FFTW), so at least the situation here is in line with the status quo.

uklotzde commented:
You could adopt our workaround in util/optional.h to enable compatibility with macOS. The pre-standard version is missing some methods from final standard and might require some workarounds.

uklotzde commented:
Please rebase on master.

uklotzde commented on CMakeLists.txt:
```diff
@@ -1438,6 +1438,71 @@ if(WIN32)
 endif()
 
+# libdjinterop (enabled only on UNIX for now)
+cmake_dependent_option(DJINTEROP "Use djinterop library" ON "UNIX;NOT APPLE" OFF)
```
This option should be disabled by default and explicitly enabled in the CI builds.

mr-smidge replied:
Changed.

uklotzde commented on CMakeLists.txt:
```diff
@@ -1438,6 +1438,71 @@ if(WIN32)
+      # Note this requires meson 0.54 or later.
+      ExternalProject_Add(libdjinterop
+        GIT_REPOSITORY ${DJINTEROP_GIT_REPO_URL}
+        GIT_TAG ${DJINTEROP_GIT_TAG}
```
Add `GIT_SHALLOW TRUE` to check out the tag with `git clone --depth 1`

mr-smidge replied:
Done.

mr-smidge commented:
> You could adopt our workaround in util/optional.h to enable compatibility with macOS. The pre-standard version is missing some methods from final standard and might require some workarounds.

@uklotzde , I thought about this approach, but wanted to check: are there plans to upgrade the version of Apple LLVM on the build servers?  If there are, I might prefer that rather than add workarounds for old compilers.

mr-smidge committed:
Make djinterop explicit opt-in.

Add -DDJINTEROP=ON to Ubuntu/CMake travis and appveyor builds.
Add shallow git clone.

mr-smidge commented:
> Please rebase on (upstream) master.

Erm, it already is (at the time of writing) - did you see something to indicate that it wasn't?  Thanks.

uklotzde commented:
> @uklotzde , I thought about this approach, but wanted to check: are there plans to upgrade the version of Apple LLVM on the build servers? If there are, I might prefer that rather than add workarounds for old compilers.

We cannot update the CI build servers as long as we want to support older macOS releases. Please add the workaround if you want your library included in macOS builds.

uklotzde commented:
> > Please rebase on (upstream) master.
> 
> Erm, it already is (at the time of writing) - did you see something to indicate that it wasn't? Thanks.

If cloned your branch locally and tried to rebase on master -> conflicts

uklotzde commented:
> > > Please rebase on (upstream) master.
> > 
> > 
> > Erm, it already is (at the time of writing) - did you see something to indicate that it wasn't? Thanks.
> 
> If cloned your branch locally and tried to rebase on master -> conflicts

Sorry, I forgot to pull master.

uklotzde commented on .travis.yml:
```diff
@@ -76,7 +76,7 @@ jobs:
       # Ubuntu Bionic build prerequisites
       # TODO for Ubuntu Focal: Replace "-DFAAD=ON" with "-DFFMPEG=ON"
-      env: CMAKEFLAGS_EXTRA="-DFAAD=ON -DLOCALECOMPARE=ON -DMAD=ON -DMODPLUG=ON -DWAVPACK=ON -DWARNINGS_FATAL=ON"
+      env: CMAKEFLAGS_EXTRA="-DDJINTEROP=ON -DFAAD=ON -DLOCALECOMPARE=ON -DMAD=ON -DMODPLUG=ON -DWAVPACK=ON -DWARNINGS_FATAL=ON"
```
This also needs to be added in appveyor.yml for the Ubuntu 20.04 build.

mr-smidge replied:
Now done.

mr-smidge commented:
> > @uklotzde , I thought about this approach, but wanted to check: are there plans to upgrade the version of Apple LLVM on the build servers? If there are, I might prefer that rather than add workarounds for old compilers.
> 
> We cannot update the CI build servers as long as we want to support older macOS releases. Please add the workaround if you want your library included in macOS builds.

Raised https://github.com/xsco/libdjinterop/issues/14 for the libdjinterop library.  I'll focus on Linux builds for this PR and will enable it for Mixxx Mac builds in a later PR.

Holzhaus commented on CMakeLists.txt:
```diff
@@ -1438,6 +1438,73 @@ if(WIN32)
+    set(DJINTEROP_GIT_REPO_URL https://github.com/xsco/libdjinterop.git)
+    set(DJINTEROP_GIT_TAG tags/0.11.0)
+    set(DJINTEROP_INSTALL_DIR "${CMAKE_CURRENT_BINARY_DIR}/lib/libdjinterop-install")
+    if(DJINTEROP_MESON_BUILD)
```
Why do we need to support both build systems? If libdjinterop has a cmake build, let's just use that one. The user might not even have meson installed and custom build flags can't be passed to meson anyway with this implementation.

mr-smidge replied:
Short bit of background: Meson has been the main build system for libdjinterop, based mainly on my personal preferences whilst developing the library, and the easier time with cross-platform dependency management.  I originally added a CMake build on a "best efforts" basis with the idea that it would make it easier to integrate with Mixxx.  As you note (and as written in a comment), the typical user compiling Mixxx may not even have Meson/Ninja installed.

So I have to think about whether I, as the libdjinterop maintainer, should support two build systems for the library.  Given the current simplicity of the libdjinterop build, maybe I can make the CMake build "official".

Nonetheless, I am not about to go and deliberately break the CMake build (!), so I don't mind removing the conditional regardless.

Holzhaus commented on CMakeLists.txt:
```diff
@@ -1438,6 +1438,73 @@ if(WIN32)
+    # builds like this one, which we will default to.
+    option(DJINTEROP_MESON_BUILD "Build djinterop library with meson" OFF)
+
+    set(DJINTEROP_GIT_REPO_URL https://github.com/xsco/libdjinterop.git)
+    set(DJINTEROP_GIT_TAG tags/0.11.0)
+    set(DJINTEROP_INSTALL_DIR "${CMAKE_CURRENT_BINARY_DIR}/lib/libdjinterop-install")
```
If we drop support for meson, these can be specified inside the `ExternalProject_Add` call.

mr-smidge replied:
Done.

mr-smidge committed:
Use only CMake build from libdjinterop

Holzhaus commented on CMakeLists.txt:
```diff
@@ -1438,6 +1438,53 @@ if(WIN32)
   target_link_libraries(mixxx-lib PUBLIC FFTW::FFTW)
 endif()
 
+# libdjinterop (must be enabled explicitly to use)
+option(DJINTEROP "Use djinterop library" OFF)
+if(DJINTEROP)
+  message(STATUS "Enabling djinterop support...")
+  target_compile_definitions(mixxx-lib PUBLIC __DJINTEROP__)
```
If possible we should expose *features*, not dependencies:

```suggestion
# Denon Engine library export support
option(DENONENGINE "Support for Denon Engine library export" OFF)
if(DENONENGINE)
  target_compile_definitions(mixxx-lib PUBLIC __DENONENGINE__)
```

mr-smidge replied:
A perfectly reasonable suggestion - I've pushed another commit to use `ENGINEPRIME` throughout (this is the naming convention we previously agreed to use in class names and expose to the user etc. in PR #2753 ).

Holzhaus commented on appveyor.yml:
```diff
@@ -115,6 +115,7 @@ for:
       -DBATTERY=ON
       -DBROADCAST=ON
       -DBULK=ON
+      -DDJINTEROP=ON
```
```suggestion
      -DDENONPRIME=ON
```

Holzhaus reviewed (changes requested):
Works fine, thanks!

mr-smidge committed:
Build options named by feature rather than dependency

Holzhaus reviewed (approved):
LGTM, thanks.

mr-smidge commented:
Many thanks @uklotzde and @Holzhaus .

The Visual Studio AppVeyor build appears to have failed tests with a large number of file-not-found errors.  I presume these are transient?  If so, please could someone kick off this job again?

mr-smidge commented:
A re-run failed too.  I might need some help with these AppVeyor VS errors - I can't see how these are connected to the commits in this PR.  Example:

```
        Start 219: EngineMicrophoneTest.TestInputMatchesOutput
219/618 Test #219: EngineMicrophoneTest.TestInputMatchesOutput .....................................***Timeout  30.01 sec
Note: Google Test filter = EngineMicrophoneTest.TestInputMatchesOutput
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from EngineMicrophoneTest
[ RUN      ] EngineMicrophoneTest.TestInputMatchesOutput
Loading resources from  "C:/projects/mixxx/cmake_build/res/"
BaseTrackPlayerImpl::slotLoadTrack "[Channel1]"
CachingReaderWorker - "[Channel1]" File not found "C:/projects/mixxx/cmake_build/src/test/sine-30.wav"
```

Any ideas?

uklotzde commented:
Strange. I'll push this branch to my personal AppVeyor account for testing.

uklotzde commented:
Test failures are probably caused by https://github.com/mixxxdj/mixxx/pull/2938

uklotzde commented:
Please merge master to verify that the build succeeds now.

mr-smidge committed:
Merge branch 'master' into enh/djinterop-external-project

mr-smidge commented:
> Please merge master to verify that the build succeeds now.

Done, and the VS AppVeyor build now passes :smile:.

Unfortunately, the OSX Scons build failed after timing out! :facepalm:  Please could this be re-run for a (hopefully) clean bill of health?

mr-smidge commented:
All green - ok to merge?

Then I'll update PR #2753 .

Holzhaus commented:
Huh? Shouldn't you merge this into #2753? This doesn't make sense in master on its own...

uklotzde commented:
@Holzhaus I think we should merge this PR before #2753 now that it is ready. It doesn't do any harm, build time are increased only slightly and the feature can easily be disabled.

Holzhaus commented:
@uklotzde okay, then do ahead

uklotzde commented:
LGTM

uklotzde commented:
Our build servers are not affected until we enable this option. Should be safe.
